session/recursive.go

```diff
@@ -0,0 +1,105 @@
+package session
+
+import (
+	"sort"
+	"strings"
+
+	"launchpad.net/gozk"
+)
+
+var defaultACL = zookeeper.ACL{
+	Perms:  zookeeper.PERM_ALL,
+	Scheme: "world",
+	Id:     "anyone",
+}
+
+// ChildrenRecursive returns a slice all of a node's descendants that are at
+// most `maxDepth` levels away from the root.
+func (s *ZKSession) ChildrenRecursive(path string, maxDepth int) ([]string, error) {
+	rootDepth := 0
+	if path != "/" {
+		rootDepth = strings.Count(path, "/")
+	}
+
+	stat, err := s.Exists(path)
+	if err != nil {
+		return nil, err
+	}
+
+	if stat == nil {
+		return []string{}, nil
+	}
+
+	nodes := []string{path}
+	for index := 0; index < len(nodes); index++ {
+		if maxDepth <= 0 || strings.Count(nodes[index], "/") < rootDepth+maxDepth {
+			if children, _, err := s.Children(nodes[index]); err == nil {
+				parent := nodes[index]
+				if parent == "/" {
+					parent = ""
+				}
+
+				for _, child := range children {
+					nodes = append(nodes, parent+"/"+child)
+				}
+			}
+		}
+	}
+	return nodes[1:], nil
+}
+
+// SetRecursive will set data for the given path, creating all parents as
+// necessary with no data.
+func (s *ZKSession) SetRecursive(path string, data string) error {
```

Comment:
I'm not 100% sold on the naming `SetRecursive` here -- my first guess was that it would take a tree of values and write them out to various keys. Maybe something like the admittedly-verbose `SetWithRecursiveCreate` or `SetWithMkdirP` or `RecursiveCreateAndSet` or `RecursiveCreateThenSet` or... I'm not sure.?